hat/backends/ptx/src/main/java/hat/backend/PTXCodeBuilder.java

```diff
@@ -0,0 +1,665 @@
+                    offset += Integer.parseInt(padding.toString());
+                }
+                case Schema.FieldNode.PrimitiveFixedArray f -> offset += f.len * 4;
+                case Schema.FieldNode.IfaceFixedArray f -> offset += f.len * 4;
+                default -> offset += 4;
+            }
+        }
+
+        if (op.fieldName.equals("array")) {
+            // append("mul.wide").s32().space().printAndAddVar(op.result(), PTXRegister.Type.U64).commaSpace()
+            //         .printVar(op.operands().getFirst()).commaSpace().append("4").ptxNl();
+            source = new PTXRegister(incrOrdinal(addrType()), addrType());
+            append("add").s64().space().append(source.name()).commaSpace().printVar(op.operands().get(0)).commaSpace().printVar(op.operands().get(1)).ptxNl();
+        } else {
+            source = getVar(op.operands().getFirst());
+        }
+
+        if (op.resultType.toString().equals("void")) {
+            st().global().u32().space().address(source.name(), offset).commaSpace().printVar(op.operands().get(2));
+        } else {
+            ld().global().u32().space().printAndAddVar(op.result(), PTXRegister.Type.U32).commaSpace().address(source.name(), offset);
+        }
+    }
+
+    public void varLoad(VarLoadOpWrapper op) {
+        ld().printResultType(op.resultType(), false).space().printResult(op, addrType()).commaSpace().printVar(op.operandNAsValue(0));
+    }
+
+    public void varStore(VarStoreOpWrapper op) {
+        st().printResultType(op.resultType(), false).space().printResult(op, addrType()).commaSpace().printVar(op.operandNAsValue(0));
+    }
+
+    public void fieldLoad(FieldLoadOpWrapper op) {
+        if (op.fieldName().equals(Field.KC_X.toString())) {
+            if (!fieldToRegMap.containsKey(Field.KC_X)) {
+                loadKcX(op.result());
+            } else {
+                mov().u32().space().printResult(op, PTXRegister.Type.U32).commaSpace().printField(Field.KC_X);
+            }
+        } else if (op.fieldName().equals(Field.KC_MAXX.toString())) {
+            if (!fieldToRegMap.containsKey(Field.KC_X)) {
+                loadKcX(op.operandNAsValue(0));
+            }
+            ld().global().u32().space().printFieldAndVal(Field.KC_MAXX, op.result()).commaSpace()
+                    .address(fieldToRegMap.get(Field.KC_ADDR).name(), 4);
+        } else {
+            ld().global().u32().space().printResult(op, PTXRegister.Type.U64).commaSpace().printVar(op.operandNAsValue(0));
+//            ld().global().u32().space().printResult(op, PTXRegister.Type.U64).commaSpace().printFieldAndVal(op.fieldName(), op.result());
+        }
+    }
+
+    public void loadKcX(Value value) {
+        append("cvta.to").global().addrSize().space().printField(Field.KC_ADDR).commaSpace()
+                .printAndAddVar(paramMap.get("kc"), addrType()).ptxNl();
+        mov().u32().space().printField(Field.NTID_X).commaSpace().append("%ntid.x").ptxNl();
```

Comment:
again 'printField ()' -> 'field()' , printAndAddVar() -> addVarAndType() .
consider 'cvta().dot().to()'
Rather than repeating these suggestions, look at all uses of printXXX and append and see it we can de